File: schedulers/mt/test/qa_parameters.cpp
#include <gtest/gtest.h>

#include <iostream>
#include <thread>

#include <gnuradio/blocklib/blocks/multiply_const.hpp>
#include <gnuradio/blocklib/blocks/throttle.hpp>
#include <gnuradio/blocklib/blocks/vector_sink.hpp>
#include <gnuradio/blocklib/blocks/vector_source.hpp>
#include <gnuradio/domain_adapter_direct.hpp>
#include <gnuradio/flowgraph.hpp>
#include <gnuradio/schedulers/mt/scheduler_mt.hpp>

using namespace gr;

TEST(SchedulerSTTest, ParameterBasic)
{
    auto src = blocks::vector_source_f::make(
        std::vector<float>{ 1.0, 2.0, 3.0, 4.0, 5.0 }, true);
    auto throttle = blocks::throttle::make(sizeof(float), 32000);
    auto mult1 = blocks::multiply_const_ff::make(100.0);
    auto snk1 = blocks::vector_sink_f::make();
    auto snk2 = blocks::vector_sink_f::make();

    flowgraph_sptr fg(new flowgraph());
    fg->connect(src, 0, throttle, 0);
    fg->connect(throttle, 0, mult1, 0);
    fg->connect(mult1, 0, snk1, 0);
    fg->connect(mult1, 0, snk2, 0);

    auto sched = schedulers::scheduler_mt::make("sched1", 4096);
    fg->set_scheduler(sched);

    fg->validate();
    fg->start();

    auto start = std::chrono::steady_clock::now();

    float k = 1.0;

    std::vector<float> k_set;
    std::vector<float> k_queried;

    while (true) {
        mult1->set_k(k);
        k_set.push_back(k);

        if (std::chrono::steady_clock::now() - start > std::chrono::seconds(1))
            break;

        std::this_thread::sleep_for(std::chrono::milliseconds(100));

        k += 1.0;

        auto query_k = mult1->k();
        // std::cout << query_k << std::endl;
        k_queried.push_back(query_k);
    }

    std::this_thread::sleep_for(std::chrono::milliseconds(1000));

    fg->stop();

    // Search for the values that we set in the values we queried
    bool all_values_found = true;
    for (unsigned i = 0; i<k_set.size()-1; i++)
    {
        if (std::find(k_queried.begin(), k_queried.end(), k_set[i]) == k_queried.end())
        {
            all_values_found = false;
        }
    }

    EXPECT_TRUE(all_values_found);

    // now look at the data
    EXPECT_GT(snk1->data().size(), 5);
    EXPECT_GT(snk2->data().size(), 5);

    // TODO - check for query
    // TODO - set changes at specific sample numbers
}

File: schedulers/nbt/test/qa_parameters.py
#!/usr/bin/env python3

from newsched import gr_unittest, gr, blocks, math
from time import time, sleep

class test_basic(gr_unittest.TestCase):

    def setUp(self):
        self.tb = gr.flowgraph()

    def tearDown(self):
        self.tb = None

    def test_basic(self):
        run_time = 1

        src = blocks.vector_source_f([1,2,3,4,5],True)
        throttle = blocks.throttle(gr.sizeof_float, 32000)
        mult1 = math.multiply_const_ff(100.0)
        snk1 = blocks.vector_sink_f()
        snk2 = blocks.vector_sink_f()

        self.tb.connect([src, throttle, mult1, snk1])
        self.tb.connect(mult1,snk2)
        
        self.tb.validate()
        self.tb.start()

        start = time()
        k = 1.0

        k_set = []
        k_queried = []

        while True: 
            mult1.set_k(k)
            k_set.append(k)

            if (time() - start > run_time):
                break

            sleep(0.1)

            k += 1.0

            query_k = mult1.k()
            # print(query_k)

            k_queried.append(query_k)


        sleep(0.05)

        self.tb.stop()

        # Search for the values that we set in the values we queried
        all_values_found = True
        for i in range(len(k_set)-1):
            if not k_set[i] in k_queried:
                all_values_found = False

        self.assertTrue(all_values_found)

        # // now look at the data
        # EXPECT_GT(snk1->data().size(), 5);
        # EXPECT_GT(snk2->data().size(), 5);

        # // TODO - check for query
        # // TODO - set changes at specific sample numbers

if __name__ == "__main__":
    gr_unittest.run(test_basic)